crates/mun_gc/src/mark_sweep.rs

```diff
@@ -0,0 +1,225 @@
+    fn default() -> Self {
+        MarkSweep {
+            objects: RwLock::new(HashMap::new()),
+            observer: O::default(),
+            stats: RwLock::new(Stats::default()),
+        }
+    }
+}
+
+impl<T: Type + Clone, O: Observer + Default> MarkSweep<T, O> {
+    pub fn new() -> Self {
+        Default::default()
+    }
+}
+
+impl<T: Type + Clone, O: Observer> MarkSweep<T, O> {
+    pub fn with_observer(observer: O) -> Self {
+        Self {
+            objects: RwLock::new(HashMap::new()),
+            observer,
+            stats: RwLock::new(Stats::default()),
+        }
+    }
+}
+
+impl<T: Type + Clone, O: Observer> MarkSweep<T, O> {
+    /// Allocates a block of memory
+    fn alloc_memory(&self, size: usize, alignment: usize) -> *mut u8 {
+        unsafe { std::alloc::alloc(Layout::from_size_align_unchecked(size, alignment)) }
+    }
+
+    /// Returns the observer
+    pub fn observer(&self) -> &O {
+        &self.observer
+    }
+}
+
+impl<T: Type + Clone, O: Observer> GCRuntime<T> for MarkSweep<T, O> {
+    fn alloc(&self, ty: T) -> GCPtr {
+        let size = ty.size();
+        let ptr = self.alloc_memory(ty.size(), ty.alignment());
+        let object = Box::pin(ObjectInfo {
+            ptr,
+            ty,
+            roots: 0,
+            color: Color::White,
+        });
+
+        // We want to return a pointer to the `ObjectInfo`, to be used as handle.
+        let handle = (object.as_ref().deref() as *const _ as RawGCPtr).into();
+
+        {
+            let mut objects = self.objects.write();
+            objects.insert(handle, object);
+        }
+
+        {
+            let mut stats = self.stats.write();
+            stats.allocated_memory += size;
+        }
+
+        self.observer.event(Event::Allocation(handle));
+        handle
+    }
+
+    unsafe fn ptr_type(&self, handle: GCPtr) -> T {
+        let _ = self.objects.read();
+
+        // Convert the handle to our internal representation
+        let object_info: *const ObjectInfo<T> = handle.into();
+
+        // Return the type of the object
+        (*object_info).ty.clone()
+    }
+
+    unsafe fn root(&self, handle: GCPtr) {
+        let _ = self.objects.write();
+
+        // Convert the handle to our internal representation
+        let object_info: *mut ObjectInfo<T> = handle.into();
+
+        // Return the type of the object
+        (*object_info).roots += 1;
+    }
+
+    unsafe fn unroot(&self, handle: GCPtr) {
+        let _ = self.objects.write();
+
+        // Convert the handle to our internal representation
+        let object_info: *mut ObjectInfo<T> = handle.into();
+
+        // Return the type of the object
+        (*object_info).roots -= 1;
+    }
+
+    fn stats(&self) -> Stats {
+        self.stats.read().clone()
+    }
+}
+
+impl<T: Type + Clone, O: Observer + Default> MarkSweep<T, O> {
+    /// Collects all memory that is no longer referenced by rooted objects. Returns `true` if memory
```

Comment:
What is your use case for knowing whether memory was reclaimed? If there isn't any, I'd remove the `bool` return type.

Reply:
For reference, in the future the bool return type will come to mean: _has finished collection_. This might be important if the GC only has a limited time window to operate and potentially still has work left to do.